docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+
+Collations are useful when comparing and sorting data. Different languages have different rules for how to order data and how to compare characters. For example, in Swedish, "ö" and "o" are in fact different letters, so they should not be equivalent in a query (`SELECT ... WHERE col LIKE '%o%'`); additionally, Swedish sorting rules place "ö" after "z" in the alphabet. Meanwhile, in German, "ö" is officially considered a separate letter from "o", but it can be sorted along with "o", expanded to sort along with "oe", or sort at the end of the alphabet, depending on the context (dictionary, phonebook, etc.). In French, diacritics never affect comparison or sorting. The issue becomes even more complex with Unicode and certain Unicode encodings (especially UTF-8) where the same logical character can be represented in many different ways, using a variable number of bytes. For example, "ö" in UTF-8 is represented by the byte sequence C3B6, but it can also be represented using the "combining diaeresis" as 6FCC88. These should compare equivalently when using a multibyte-aware collation. Implementing this functionality as far down as possible in the database reduces the number of rows that must be shipped around.
+
+In MySQL, collation is treated as an attribute of the character set: for each character set has options for different collations(and one of them is the default collation), every collation belongs to one character set only. The effects of collation are:
+  * It defines a total order on all characters of a character set.
+  * It determines if padding is applied when comparing two strings.
+  * It describes comparisons of logically identical but physically different byte sequences.
+Some examples of the effects of collation can be found in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-effect.html) of the MySQL Manual and [this webpage](http://demo.icu-project.org/icu-bin/locexp?_=en_US&x=col).
+
+Several collation implementations are described in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-implementations.html) of MySQL Manual:
+  * Per-character table lookup for 8-bit character sets, for example: `latin1_swedish_ci`.
+  * Per-character table lookup for multibyte character sets, including some of the Unicode collations, for example: `utf8mb4_general_ci`.
+    - Full UCA implementation for some Unicode collations, for example: `utf8mb4_900_ai_ci`.
+
+## Proposal
+
+### Collate Interface
+
+There are two basic functions needed for collation `sortKey` calculation:
+  * Function that is used to compare two strings according to the given collation.
+  * Function that is used to make a memory-comparable `sortKey` corresponding to the given string.
+The interface can be defined as following:
+
+```
+type CollAttrFlag int
+
+const (
+    FlagIgnoreCase uint64=1 << iota
+    FlagIgnoreAccent
+    FlagPadding
```

Comment:
If you don't ignore accent, what is the defined order of the accents - byte order? The sorting order may be defined (for example Canadian French and French sort accents in the opposite order.. but there is likely a more useful example where this matters).

UCA implements this by having a Character Weight, Accent Weight, Case Weight, and Punctuation Weight.

Reply:
> If you don't ignore accent, what is the defined order of the accents - byte order? 

I think the order can be defined by the raw bytes/codepoint/accent weight of UCA, it depends on the demands.